logic/src/commonMain/kotlin/com/wire/kalium/logic/data/event/EventRepository.kt

```diff
@@ -0,0 +1,88 @@
+package com.wire.kalium.logic.data.event
+
+import com.wire.kalium.logic.CoreFailure
+import com.wire.kalium.logic.data.client.ClientRepository
+import com.wire.kalium.logic.data.conversation.ClientId
+import com.wire.kalium.logic.functional.Either
+import com.wire.kalium.logic.functional.suspending
+import com.wire.kalium.network.api.notification.NotificationApi
+import com.wire.kalium.network.api.notification.NotificationPageResponse
+import com.wire.kalium.network.utils.NetworkResponse
+import com.wire.kalium.network.utils.isSuccessful
+import com.wire.kalium.persistence.event.EventInfoStorage
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.flow.flattenConcat
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.isActive
+import kotlin.coroutines.coroutineContext
+
+interface EventRepository {
+    suspend fun events(): Flow<Either<CoreFailure, Event>>
+}
+
+class EventDataSource(
+    private val notificationApi: NotificationApi,
+    private val eventInfoStorage: EventInfoStorage,
+    private val clientRepository: ClientRepository,
+    private val eventMapper: EventMapper
+) : EventRepository {
+
+    override suspend fun events(): Flow<Either<CoreFailure, Event>> = suspending {
+        clientRepository.currentClientId().coFold({
+            flowOf(Either.Left(CoreFailure.MissingClientRegistration))
+        }, { clientId ->
+            val pendingEventsFlow = pendingEventsFlow(clientId)
+            val liveEventsFlow = liveEventsFlow(clientId)
+            flowOf(pendingEventsFlow, liveEventsFlow).flattenConcat()
+        })!!
```

Comment:
Wouldn't this result be always non-nullable?

Reply:
Yes. I was based on the old implementation of `fold` and `coFold`. I can remove the double-bang now.